### PR TITLE
shaohaos lessons module for mw and fix rdruid 2pc

### PR DESCRIPTION
**Pull request by trevorm4**
Adding shaohaos module
### Testing

image
image

### CODE REVIEW

File: src/analysis/retail/monk/mistweaver/modules/spells/ShaohaosLessons.tsx [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+    if (!this.selectedCombatant.hasBuff(SPELLS.LESSON_OF_DESPAIR_BUFF.id)) {
+      return;
+    }
+    if (this.curHpPercent < 0.5 || event.hitType !== HIT_TYPES.CRIT) {
+      return;
+    }
+    const critAmount = event.amount / 2; // crit = 200% heal
+    const critIncrease =
+      DESPAIR_CRIT_INCREASE / (this.statTracker.currentCritPercentage + DESPAIR_CRIT_INCREASE);
+    this.despairHealing += critAmount * critIncrease;
```
Comment (vohrr) [inline src/analysis/retail/monk/mistweaver/modules/spells/ShaohaosLessons.tsx]: need to get the effective healing from these events

File: src/analysis/retail/monk/mistweaver/modules/spells/ShaohaosLessons.tsx [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+      return;
+    }
+    if (this.curHpPercent < 0.5 || event.hitType !== HIT_TYPES.CRIT) {
+      return;
+    }
+    const critAmount = event.amount / 2; // crit = 200% heal
+    const critIncrease =
+      DESPAIR_CRIT_INCREASE / (this.statTracker.currentCritPercentage + DESPAIR_CRIT_INCREASE);
+    if (event.type === EventType.Heal) {
+      this.despairHealing += critAmount * critIncrease;
```
Comment (vohrr) [inline src/analysis/retail/monk/mistweaver/modules/spells/ShaohaosLessons.tsx]: Effective healing and effective dmg to these

File: src/analysis/retail/monk/mistweaver/modules/spells/ShaohaosLessons.tsx [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+    );
+    this.addEventListener(Events.heal.by(SELECTED_PLAYER), this.handleHealDoubt);
+    this.addEventListener(Events.damage.by(SELECTED_PLAYER), this.handleDamageDoubt);
+    this.addEventListener(Events.heal.by(SELECTED_PLAYER), this.handleDespair);
+    this.addEventListener(Events.damage.by(SELECTED_PLAYER), this.handleDespair);
+    this.addEventListener(Events.heal.to(SELECTED_PLAYER), this.healTaken);
+    this.addEventListener(Events.damage.to(SELECTED_PLAYER), this.damageTaken);
+  }
+
+  healTaken(event: HealEvent) {
```
Comment (vohrr) [inline src/analysis/retail/monk/mistweaver/modules/spells/ShaohaosLessons.tsx:65]: You can also  add a cast event listener to increase the accuracy of health tracking here

**Review by vohrr — CHANGES_REQUESTED** [review]

**Comment by vohrr** [comment]
Just the two and then looks good

**Review by kfinch — APPROVED** [review]
Resto adjustment seems fine to me

**Review by vohrr — APPROVED** [review]

**Pull request merged**